File: android_app/app/src/main/java/com/rappipay_demo/android_app/multiplatform/MultiplatformDemoEngine.kt
package com.rappipay_demo.android_app.multiplatform

import android.content.Context
import com.rappipay_demo.android_app.multiplatform.exceptions.EngineAlreadyExistsException
import com.rappipay_demo.android_app.multiplatform.plugin.MultiplatformDemoPlugin
import io.flutter.embedding.android.FlutterFragment
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.embedding.engine.FlutterEngineCache
import io.flutter.embedding.engine.dart.DartExecutor

class MultiplatformDemoEngine(private val engineId: String, private val flutterEngine: FlutterEngine) {
    constructor(context: Context, engineId: String) : this(
        engineId, createFlutterEngine(
            context,
            engineId
        )
    )

    private val currentMultiplatformDemoPlugin: MultiplatformDemoPlugin? by lazy {
        flutterEngine.plugins?.get(MultiplatformDemoPlugin::class.java) as MultiplatformDemoPlugin?
    }

    val router
        get() = currentMultiplatformDemoPlugin?.channelChannel

    fun buildNewComponent() : MultiplatformDemoComponent {
        val flutterFragment : FlutterFragment = FlutterFragment.withCachedEngine(engineId).build()
        return MultiplatformDemoComponent(flutterFragment)
    }

    fun stop() {
        flutterEngine.destroy()
        FlutterEngineCache.getInstance().remove(engineId)
        MultiplatformDemo.aliveEngines.remove(engineId)
    }

    companion object {
        private fun createFlutterEngine(context: Context, engineId: String) : FlutterEngine {
            if (instanceExists(engineId)) {
                throw EngineAlreadyExistsException(engineId)
            }

            val newFlutterEngine = FlutterEngine(context)

            newFlutterEngine.dartExecutor.executeDartEntrypoint(
                DartExecutor.DartEntrypoint.createDefault()
            )

            FlutterEngineCache
                .getInstance()
                .put(engineId, newFlutterEngine)

            MultiplatformDemo.aliveEngines.add(engineId)

            newFlutterEngine.plugins.add(MultiplatformDemoPlugin())

            return newFlutterEngine
        }

        fun getInstance(engineId: String) : MultiplatformDemoEngine? {
            return FlutterEngineCache.getInstance().get(engineId)?.let { flutterEngine ->
                MultiplatformDemoEngine(engineId, flutterEngine)
            }
        }

        fun instanceExists(engineId: String) = FlutterEngineCache.getInstance().contains(engineId)
    }
}

File: ios_app/iosApp/Multiplatform/MultiplatformDemoEngine.swift
//
//  MultiplatformDemoEngine.swift
//  iosApp
//
//  Created by Liber Da Silva on 01/12/2020.
//

import Foundation
import Flutter

open class MultiplatformDemoEngine {
    private let engineId: String
    private let flutterEngine: FlutterEngine
    private let multiplatformDemoPlugin : MultiplatformDemoPlugin
    
    public lazy var router: RouterChannel = multiplatformDemoPlugin.routerChannel

    init(_ engineId: String) throws {
        let newFlutterEngine = try MultiplatformDemoEngine.createFlutterEngine(engineId)
        
        self.engineId = engineId
        self.flutterEngine = newFlutterEngine
        self.multiplatformDemoPlugin = MultiplatformDemoPlugin(
            with: flutterEngine.registrar(
                forPlugin: String(describing: MultiplatformDemoPlugin.self)
            )
        )
        
        MultiplatformDemo.instance.aliveEngines[engineId] = self
    }
    
    public func buildNewComponent() -> MultiplatformDemoComponent {
        let flutterViewController = FlutterViewController(engine: flutterEngine, nibName: nil, bundle: nil)
        return MultiplatformDemoComponent(flutterViewController)
    }
    
    public func stop() {
        multiplatformDemoPlugin.destroy()
        flutterEngine.destroyContext()
        MultiplatformDemo.instance.aliveEngines.removeValue(forKey: engineId)
    }

    private static func createFlutterEngine(_ engineId: String) throws -> FlutterEngine {
        if (MultiplatformDemo.instance.engineInstanceExists(engineId)) {
            throw MultiplatformDemoEngineError.engineAlreadyExists(engineId)
        }
        
        let engine = FlutterEngine(name: engineId)
        engine.run()
                
        return engine
    }
}
